src/lib.rs

```diff
@@ -204,59 +250,224 @@ pub fn encode_mode(bytes: &[u8], mode: Base64Mode) -> String {
+                bad_byte_index = input_index + 3;
+                break;
+            };
+            accum |= (morsel as u64) << 40;
+
+            morsel = decode_table[(input_chunk >> 24 & 0xFF) as usize];
+            if morsel == decode_tables::INVALID_VALUE {
+                bad_byte_index = input_index + 4;
+                break;
+            };
+            accum |= (morsel as u64) << 34;
+
+            morsel = decode_table[(input_chunk >> 16 & 0xFF) as usize];
+            if morsel == decode_tables::INVALID_VALUE {
+                bad_byte_index = input_index + 5;
+                break;
+            };
+            accum |= (morsel as u64) << 28;
+
+            morsel = decode_table[(input_chunk >> 16 & 0xFF) as usize];
+            if morsel == decode_tables::INVALID_VALUE {
+                bad_byte_index = input_index + 6;
+                break;
+            };
+            accum |= (morsel as u64) << 22;
+
+            morsel = decode_table[(input_chunk & 0xFF) as usize];
+            if morsel == decode_tables::INVALID_VALUE {
+                bad_byte_index = input_index + 7;
+                break;
+            };
+            accum |= (morsel as u64) << 16;
+
+            BigEndian::write_u64(&mut buffer_slice[(output_index)..(output_index + 8)],
+                                 accum);
+
+            output_index += 6;
+            input_index += chunk_len;
+        };
+
+        if morsel == decode_tables::INVALID_VALUE {
+            // we got here from a break
+            return Err(Base64Error::InvalidByte(bad_byte_index, input_bytes[bad_byte_index]));
+        }
     }
 
-    let div = buffer.len() - rem;
-
-    let mut raw = Vec::<u8>::with_capacity(3*div/4 + rem);
-    let mut i = 0;
+    // Truncate off the last two bytes from writing the last u64.
+    // Unconditional because we added on the extra 2 bytes in the resize before the loop.
+    let new_len = buffer.len() - 2;
+    buffer.truncate(new_len);
+
+    // handle leftovers (at most 8 bytes).
+    // Use a u64 as a stack-resident 8-byte Vec.
+    let mut leftover_bits: u64 = 0;
+    let mut morsels_in_leftover = 0;
+    let mut padding_bytes = 0;
+    let mut first_padding_index: usize = 0;
+    for (i, b) in input.as_bytes()[length_of_full_chunks..].iter().enumerate() {
+        // '=' padding
+        if *b == 0x3D {
+            // There can be bad padding in a few ways:
+            // 1 - Padding with non-padding characters after it
+            // 2 - Padding after zero or one non-padding characters before it
+            //     in the current quad.
+            // 3 - More than two characters of padding. If 3 or 4 padding chars
+            //     are in the same quad, that implies it will be caught by #2.
+            //     If it spreads from one quad to another, it will be caught by
+            //     #2 in the second quad.
+
+            if i % 4 < 2 {
+                // Check for case #2.
+                // TODO InvalidPadding error
```

Comment:
Something to ponder: This isn't really an "invalid byte" in the since of illegal input to decode, it's that padding is wrong. Should we have a separate `InvalidPadding`? Pro: we *can* detect this error. Con: Probably nobody will ever care. 

I'm leaning towards "don't bother" at the moment but maybe this will solve somebody's longstanding problem!